src/jdk.jdeps/share/classes/com/sun/tools/jnativescan/JNativeScanTask.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.tools.jnativescan;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.lang.constant.ClassDesc;
+import java.lang.module.Configuration;
+import java.lang.module.ModuleFinder;
+import java.lang.module.ResolvedModule;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import java.util.zip.ZipFile;
+
+class JNativeScanTask {
+
+    private final PrintWriter out;
+    private final List<Path> classPaths;
+    private final List<Path> modulePaths;
+    private final List<String> cmdRootModules;
+    private final Runtime.Version version;
+    private final Action action;
+
+    public JNativeScanTask(PrintWriter out, List<Path> classPaths, List<Path> modulePaths,
+                           List<String> cmdRootModules, Runtime.Version version, Action action) {
+        this.out = out;
+        this.classPaths = classPaths;
+        this.modulePaths = modulePaths;
+        this.version = version;
+        this.action = action;
+        this.cmdRootModules = cmdRootModules;
+    }
+
+    public void run() throws JNativeScanFatalError {
+        List<ClassFileSource> toScan = new ArrayList<>(findAllClassPathJars());
+
+        ModuleFinder moduleFinder = ModuleFinder.of(modulePaths.toArray(Path[]::new));
+        List<String> rootModules = cmdRootModules;
+        if (rootModules.contains("ALL-MODULE-PATH")) {
+            rootModules = allModuleNames(moduleFinder);
+        }
+        Configuration config = Configuration.resolveAndBind(moduleFinder, List.of(systemConfiguration()),
+                ModuleFinder.of(), rootModules);
```

Comment:
I think the module path should work like other tools so the "moduleFinder" should be "after" finder, not the "before" finder. In other words, the modules that are observable on the application module path don't override the system modules. If you want you can use an instance method here, like this:

`systemConfiguration().resovleAndBind(ModuleFinder.of(), moduleFinder, rootModules)`